### PR TITLE
ps kernel hard code handling changes

### PR DESCRIPTION
**Pull request by karthdmg-xilinx**
#### Problem solved by the commit
The existing code reads/assumes ps kernel slot at 0th index by default when loading the xclbin based on UUID.
Made changes such that User queries based on UUID using IOCTL and reads the appropriate AXLF buffer.
#### Bug / issue (if any) fixed, which PR introduced the bug, how it was discovered
Program PS kernel on latest XRT.
#### How problem was solved, alternative solutions (if any) and why they were rejected
eliminate hard codes changes of loading ps kernel.
#### Risks (if any) associated the changes in the commit
None
#### What has been tested and how, request additional testing if necessary
Program PS, PL/PS and AIE kernels and verify the behavior.
#### Documentation impact (if any)
Need to update multi slot behavior for ps kernels.

### CODE REVIEW

File: src/runtime_src/core/common/device.cpp [hunk 1]
```diff
@@ -146,20 +146,44 @@ load_xclbin(const uuid& xclbin_id)
...
+  xrt_core::query::aie_skd_xclbin::args arg = {reinterpret_cast<uint64_t>(ps_uuid_ptr), reinterpret_cast<uint64_t>(skd_axlf_ptr)};
+
+  try {
+      xrt_core::device_query<xrt_core::query::aie_skd_xclbin>(this, arg);
+  }
+  catch (const std::exception &e) {
+    std::cerr << boost::format("ERROR: Failed to get xclbin for uuid (%lx)\n %s\n") % uuid_loaded.get() % e.what();
+    throw xrt_core::error(std::errc::operation_canceled);
+  }
+#endif
```
Comment (stsoe) [inline src/runtime_src/core/common/device.cpp:173]: This needs commenting and cannot use `new`.

File: src/runtime_src/core/include/xclbin.h [hunk 2]
```diff
@@ -108,6 +108,8 @@ typedef GUID xuid_t;
 #define SIZE_OF_STRUCT(s) \
    char (*__fail)[sizeof(struct s)] = 1
 
+#define MAX_AXLF_BUF_SIZE       (4096*1024)
```
Comment (stsoe) [inline src/runtime_src/core/include/xclbin.h:111]: Looks random.   What is this and why?

File: src/runtime_src/core/common/query_requests.h [hunk 3]
```diff
@@ -3819,6 +3820,23 @@ struct read_trace_data : request
   virtual std::any
   get(const device*, const std::any&) const = 0;
 };
+
+struct aie_skd_xclbin : request
```
Comment (stsoe) [inline src/runtime_src/core/common/query_requests.h:3824]: Add a comment about this request type.

**Review by stsoe — CHANGES_REQUESTED** [review]

**Review by chvamshi-xilinx — CHANGES_REQUESTED** [review]
As discussed offline, Please simplify the logic and update existing load_xclbin to support this case also.
No new #ifdefs . We are already struggling to remove AIE* defines

File: src/runtime_src/core/include/xclbin.h (hunk 2 above)
Comment (karthdmg-xilinx) [reply src/runtime_src/core/include/xclbin.h:111]: Removed

File: src/runtime_src/core/common/device.cpp (hunk 1 above)
Comment (karthdmg-xilinx) [reply src/runtime_src/core/common/device.cpp]: Added another query to read axlf size, please review

File: src/runtime_src/core/common/query_requests.h (hunk 3 above)
Comment (karthdmg-xilinx) [reply src/runtime_src/core/common/query_requests.h:3826]: Added

File: src/runtime_src/core/edge/drm/zocl/common/zocl_sysfs.c [hunk 4]
```diff
@@ -600,7 +600,7 @@ static ssize_t read_xclbin_full(struct file *filp, struct kobject *kobj,
 	read_lock(&zdev->attr_rwlock);
 
 	// Only read slot 0's xclbin - TODO: extend to multi-slot 
-	zocl_slot = zdev->pr_slot[0];
+	zocl_slot = zdev->pr_slot[1];
```
Comment (saifuddin-xilinx) [inline src/runtime_src/core/edge/drm/zocl/common/zocl_sysfs.c:603]: This index should come as an input. 
Think about, if two or more PS XCLBINs are  loaded

File: src/runtime_src/core/edge/drm/zocl/edge/zocl_aie.c [hunk 5]
```diff
@@ -552,6 +553,98 @@ int zocl_aie_freqscale(struct drm_zocl_dev *zdev, void *data)
...
+		return -ENOMEM;
+
+	 ret = copy_from_user(skd_uuid, uuid_ptr, sizeof(uuid_t));
+	if (ret) {
+		vfree(skd_uuid);
+		return ret;
+	}
+	mutex_lock(&zdev->aie_lock);
+
+	for (i = 0;i < MAX_PR_SLOT_NUM;i++) {
```
Comment (saifuddin-xilinx) [inline src/runtime_src/core/edge/drm/zocl/edge/zocl_aie.c:576]: Please correct the format 
for (i = 0; i < MAX_PR_SLOT_NUM; i++)

File: src/runtime_src/core/edge/drm/zocl/edge/zocl_aie.c [hunk 6]
```diff
@@ -552,6 +553,98 @@ int zocl_aie_freqscale(struct drm_zocl_dev *zdev, void *data)
...
+		struct drm_zocl_slot *slot = NULL;
+		slot = zdev->pr_slot[i];
```
Comment (saifuddin-xilinx) [inline src/runtime_src/core/edge/drm/zocl/edge/zocl_aie.c:578]: Sanity Check required, before accessing the fields

File: src/runtime_src/core/edge/drm/zocl/edge/zocl_aie.c [hunk 7]
```diff
@@ -552,6 +553,98 @@ int zocl_aie_freqscale(struct drm_zocl_dev *zdev, void *data)
...
+		return -ENOMEM;
+
+	 ret = copy_from_user(skd_uuid, uuid_ptr, sizeof(uuid_t));
+	if (ret) {
+		vfree(skd_uuid);
+		return ret;
+	}
+	mutex_lock(&zdev->aie_lock);
+
+	for (i = 0;i < MAX_PR_SLOT_NUM;i++) {
```
Comment (saifuddin-xilinx) [inline src/runtime_src/core/edge/drm/zocl/edge/zocl_aie.c:622]: Please correct the format

File: src/runtime_src/core/edge/drm/zocl/edge/zocl_aie.c [hunk 8]
```diff
@@ -552,6 +553,98 @@ int zocl_aie_freqscale(struct drm_zocl_dev *zdev, void *data)
...
+		struct drm_zocl_slot *slot = NULL;
+		slot = zdev->pr_slot[i];
+		mutex_lock(&zdev->pr_slot[i]->slot_xclbin_lock);
+		slot_uuid = zocl_xclbin_get_uuid(zdev->pr_slot[i]);
```
Comment (saifuddin-xilinx) [inline src/runtime_src/core/edge/drm/zocl/edge/zocl_aie.c:626]: Sanity check for unused slot

File: src/runtime_src/core/edge/drm/zocl/edge/zocl_aie.c [hunk 9]
```diff
@@ -552,6 +553,98 @@ int zocl_aie_freqscale(struct drm_zocl_dev *zdev, void *data)
...
+				mutex_unlock(&zdev->pr_slot[i]->slot_xclbin_lock);
+				break;
+			}
+		}
+		mutex_unlock(&zdev->pr_slot[i]->slot_xclbin_lock);
+	}
+	if(i == MAX_PR_SLOT_NUM)
+	{
+		DRM_ERROR(" %s: UUID not found \n",__func__);
+		return -ENODEV;
```
Comment (saifuddin-xilinx) [inline src/runtime_src/core/edge/drm/zocl/edge/zocl_aie.c:639]: mutex_unlock(&zdev->aie_lock);

**Review by saifuddin-xilinx — CHANGES_REQUESTED** [review]

File: src/runtime_src/core/edge/drm/zocl/common/zocl_sysfs.c (hunk 4 above)
Comment (karthdmg-xilinx) [reply src/runtime_src/core/edge/drm/zocl/common/zocl_sysfs.c:603]: this is a dead code, sysfs functionality is being used anymore

File: src/runtime_src/core/include/xclbin.h (hunk 2 above)
Comment (chvamshi-xilinx) [reply src/runtime_src/core/include/xclbin.h:111]: i Still see this macro

File: src/runtime_src/core/edge/skd/xrt_skd.cpp [hunk 10]
```diff
@@ -52,6 +52,7 @@ namespace xrt {
       m_parent_bo_handle(xrt::shim_int::get_buffer_handle(xrtDeviceToXclDevice(handle), parent_mem_bo_in)),
       m_mem_start_paddr(mem_start_paddr_in),
       m_mem_size(mem_size_in)
+
```
Comment (chvamshi-xilinx) [inline src/runtime_src/core/edge/skd/xrt_skd.cpp:55]: remove the new line?

File: src/runtime_src/core/edge/user/device_linux.cpp [hunk 11]
```diff
@@ -841,6 +842,65 @@ struct read_trace_data
...
+#if defined(XRT_ENABLE_AIE)
+    struct drm_zocl_skd_axlf_size aie_arg;
+    aie_arg.ps_uuid_ptr = std::any_cast<xrt_core::query::skd_axlf_size::ps_uuid_ptr_type>(skd_uuid_ptr);
+
+    const std::string zocl_device = "/dev/dri/" + get_render_devname();
+    auto fd_obj = aie_get_drmfd(device, zocl_device);
+    if (fd_obj->fd < 0)
+      throw xrt_core::error(-EINVAL, boost::str(boost::format("Cannot open %s") % zocl_device));
+
+    if (ioctl(fd_obj->fd, DRM_IOCTL_ZOCL_SKD_AXLF_SIZE, &aie_arg))
```
Comment (chvamshi-xilinx) [inline src/runtime_src/core/edge/user/device_linux.cpp:894]: please dont call ioctls from here. create shim function and call it..

File: src/runtime_src/core/edge/user/device_linux.cpp [hunk 12]
```diff
@@ -841,6 +842,65 @@ struct read_trace_data
...
+  XAie_DevInst* devInst;         // AIE Device Instance
+
+  static result_type
+  get(const xrt_core::device* device, key_type key, const std::any& skd_uuid_ptr)
+  {
+#if defined(XRT_ENABLE_AIE)
+    struct drm_zocl_skd_axlf_size aie_arg;
+    aie_arg.ps_uuid_ptr = std::any_cast<xrt_core::query::skd_axlf_size::ps_uuid_ptr_type>(skd_uuid_ptr);
+
+    const std::string zocl_device = "/dev/dri/" + get_render_devname();
```
Comment (chvamshi-xilinx) [inline src/runtime_src/core/edge/user/device_linux.cpp:889]: shim function should take care of this. please dont repeat the logic

File: src/runtime_src/core/edge/user/device_linux.cpp [hunk 13]
```diff
@@ -841,6 +842,65 @@ struct read_trace_data
...
+  get(const xrt_core::device* device, key_type key, const std::any& param)
+  {
+#if defined(XRT_ENABLE_AIE)
+    struct drm_zocl_aie_skd_xclbin aie_arg;
+    auto args = std::any_cast<xrt_core::query::aie_skd_xclbin::args>(param);
+
+    aie_arg.ps_uuid_ptr = std::any_cast<xrt_core::query::aie_skd_xclbin::ps_uuid_ptr_type>(args.ps_uuid_ptr);
+    aie_arg.skd_axlf_ptr = std::any_cast<xrt_core::query::aie_skd_xclbin::skd_axlf_ptr_type>(args.skd_axlf_ptr);
+
+    const std::string zocl_device = "/dev/dri/" + get_render_devname();
```
Comment (chvamshi-xilinx) [inline src/runtime_src/core/edge/user/device_linux.cpp:861]: same as below. create shim function to handle this

File: src/runtime_src/core/edge/include/zynq_ioctl.h [hunk 14]
```diff
@@ -124,7 +124,9 @@ enum drm_zocl_ops {
 	DRM_ZOCL_AIE_FREQSCALE,
 	/* Set CU read-only range */
 	DRM_ZOCL_SET_CU_READONLY_RANGE,
-	DRM_ZOCL_NUM_IOCTLS
+	DRM_ZOCL_NUM_IOCTLS,
```
Comment (chvamshi-xilinx) [inline src/runtime_src/core/edge/include/zynq_ioctl.h:127]: NUM_IOCTL should be at the end. it should be last macro.

File: src/runtime_src/core/common/query_requests.h [hunk 15]
```diff
@@ -3819,6 +3821,35 @@ struct read_trace_data : request
...
+  struct args {
+    uint64_t  ps_uuid_ptr;
+    uint64_t  skd_axlf_ptr;
+  };
+
+  virtual std::any
+  get(const device*, const std::any& arg) const = 0;
+};
+
+struct skd_axlf_size : request
```
Comment (chvamshi-xilinx) [inline src/runtime_src/core/common/query_requests.h:3842]: add comment for this too.

File: src/runtime_src/core/common/device.cpp [hunk 16]
```diff
@@ -146,14 +145,39 @@ load_xclbin(const uuid& xclbin_id)
...
+  if(axlf_size == 0)
+  {
+    std::cerr << boost::format("ERROR: UUID not found (%lx)\n %s\n") % uuid_loaded.get();
+    throw xrt_core::error(std::errc::operation_canceled);
+  }
+  /* Initialize the axlf buffer and query axlf for the specified UUID */
+  std::vector<char> buf(axlf_size, 0);
 
-  const axlf* top = reinterpret_cast<axlf *>(xclbin_full.data());
+  xrt_core::query::aie_skd_xclbin::args arg = {reinterpret_cast<uint64_t>(ps_uuid_ptr), reinterpret_cast<uint64_t>(buf.data())};
```
Comment (chvamshi-xilinx) [inline src/runtime_src/core/common/device.cpp:168]: ps_uuid_ptr is already uint64_t, dont call reinterpret cast.

**Review by chvamshi-xilinx — CHANGES_REQUESTED** [review]

File: src/runtime_src/core/edge/drm/zocl/edge/zocl_aie.c [hunk 17]
```diff
@@ -552,6 +553,99 @@ int zocl_aie_freqscale(struct drm_zocl_dev *zdev, void *data)
...
+	 ret = copy_from_user(skd_uuid, uuid_ptr, sizeof(uuid_t));
+	if (ret) {
+		vfree(skd_uuid);
+		return ret;
+	}
+	mutex_lock(&zdev->aie_lock);
+
+	for (i = 0; i < MAX_PR_SLOT_NUM; i++) {
+		struct drm_zocl_slot *slot = NULL;
+		slot = zdev->pr_slot[i];
```
Comment (saifuddin-xilinx) [inline src/runtime_src/core/edge/drm/zocl/edge/zocl_aie.c:624]: if (!slot)
        continue;

File: src/runtime_src/core/edge/drm/zocl/edge/zocl_aie.c (hunk 6 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/edge/drm/zocl/edge/zocl_aie.c:578]: if (!slot)
        continue;

**Comment by karthdmg-xilinx** [comment]
Not handling the multi slot changes for now, hence rejecting this PR

**Pull request closed without merging**